### PR TITLE
Initial commit. Includes basic feature set without documentation

### PR DESCRIPTION
**Pull request by sriraamas**
Removed values in core/config.php which is sensitive.
Basic commit includes following features.

User Self Service : 
Change password
Generate SSH Credentials
Generate VPN Credentials

Admin Self service:
Reset Password
Revoke VPN and SSH Access
Create New User
Search Users/Groups

### CODE REVIEW

File: admin/ldapss.log [hunk 1]
```diff
@@ -0,0 +1,262 @@
+Admin Object Initialized
```
Comment (krjackso) [inline admin/ldapss.log]: You should add *.log to the .gitignore

File: admin/newUser.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+			if ($status === true){
+				$response = array('success' => true, 'data' => array("username" => $uname, "password" => $pwd) ,'errors' => array() );
+			} else {
+				$response = array('success' => false, 'errors' => array($status));
+			}
+		} else {
+				throw new Exception("UserName Already Exists. Please Choose another username");
+		}
+	} else {
+		throw new Exception("No Admin Credentials Received");
```
Comment (krjackso) [inline admin/newUser.php]: I see you throwing an Exception only to catch it right away.  I think it's appropriate to catch exceptions but throwing them doesn't gain us anything but make it slightly harder to see what the response is when we are in that case.

File: admin/newUser.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	}
+}
+catch (Exception $e) {
+		$response = array('success' => false, 'errors' => array($e->getMessage()) );
+}
+header('Content-Type: application/json',TRUE);
+$jsonstring = json_encode($response);
+echo $jsonstring;
+?>
```
Comment (krjackso) [inline admin/newUser.php]: Not 100% sure but I think we are supposed to not have the trailing `?>` in php.  @matthew-lucidchart would know for sure

File: user-vpn-ss.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+<script src="js/util.js"></script>
+<script src="js/form.js"></script>
+<script src="js/user.js"></script>
+<script>
+      $(document).foundation({
+        abide: {
+              live_validate : true, // validate the form as you go
+              error_labels: true, // labels with a for="inputId" will recieve an `error` class
+              validators :{
+                isStrong: function(el,required,parent){
```
Comment (krjackso) [inline user-vpn-ss.php]: A little nit-picky but the whitespace on these 4 lines duplicated on 20 files is really annoying :p

File: core/util.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+function LDAPtoTS($ldap_ts) {
+	$ts = new DateTime();
+    $secsAfterADepoch = $ldap_ts / 10000000;
+    $unix_ts = intval( $secsAfterADepoch - 11644473600.0);
+   	$ts -> setTimestamp($unix_ts);
+    return $ts;
+}
+
+function generate_ssl_keypair($username, $key_length){
```
Comment (krjackso) [inline core/util.php]: This function looks pretty sloppy.  In most other places we use camelCase for the function names.  We should make sure to do it the same in all the php files.  I also see some weird white space down here that stands out a lot, especially in the `$dn` array.

File: core/user.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        $zip_filename = "$this->username.vpn.credentials.zip";
+        $status = $zip -> open($zip_filename,ZipArchive::OVERWRITE);
+        if ($status!== TRUE){
+            throw new Exception("Cannot create Zip File");
+        }
+        list ($cert,$pub,$priv) = generate_ssl_keypair($this->username,2048);
+        $zip -> addFromString("client.key", $priv);
+        $zip -> addFromString("client.crt", $cert);
+        $status = $zip -> close();
+        $update_status = $this -> updateMyAttribute("lucidVpnCert", $pub);
```
Comment (krjackso) [inline core/user.php]: Sometimes we use underscores, and sometimes we use camelCase for these variables.  We should pick one and stick to it.

File: admin/check_username.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<?php
+require('../core/admin.php');
+try {
+    $admin_uname = trim($_REQUEST['aname']);
```
Comment (matthew-lucidchart) [inline admin/check_username.php]: use POST instead of REQUEST

File: admin/check_username.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    } else {
+        throw new Exception("No Admin Credentials Received");
+    }
+} catch (Exception $e){
+    $response = array('success' => false, 'errors' => array($e->getMessage()) );
+}
+header('Content-Type: application/json',TRUE);
+$jsonstring = json_encode($response);
+echo $jsonstring;
+?>
```
Comment (matthew-lucidchart) [inline admin/check_username.php]: remove all closing PHP tags

File: admin/check_username.php (hunk 7 above)
Comment (matthew-lucidchart) [inline admin/check_username.php]: Missing XCRF token

File: admin/newUser.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
+<?php
+require('../core/admin.php');
+try {
```
Comment (matthew-lucidchart) [inline admin/newUser.php]: try's are misleading. only $userObj->createUser will throw an exception

File: admin/search.php [hunk 10]
```diff
@@ -0,0 +1,73 @@
+<?php
+require('../core/admin.php');
```
Comment (matthew-lucidchart) [inline admin/search.php]: require_once

File: admin/search.php [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+try{
+
+    $aname = trim($_REQUEST['aname']);
+    $apwd = trim($_REQUEST['apwd']);
+    if (!empty($aname) && !empty($apwd)){
+        $userObj = new Admin($aname, $apwd);
+        $data = array();
+        if(!empty($_REQUEST['groupQuery'])){
+            $group = trim($_REQUEST['groupQuery']);
+            $filter = "(|(sAMAccountName=$group*)(sn=$group*)(givenName=$group*))";
```
Comment (matthew-lucidchart) [inline admin/search.php]: should $group be escaped here?

File: core/config.php [hunk 12]
```diff
@@ -0,0 +1,8 @@
+[main]
```
Comment (matthew-lucidchart) [inline core/config.php]: name? config.ini?

File: core/logger.php [hunk 13]
```diff
@@ -0,0 +1,17 @@
+<?php
+class Logger{
+    public $filename = "ldapss.log";
+    public $fp;
+    public function __construct(){
+        $this -> fp = fopen($this -> filename,"a+");
```
Comment (matthew-lucidchart) [inline core/logger.php]: this gets a lock. it means only one request that uses logging can be executed at a time

File: core/lucid_ldap.php [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+    public function getUsersinGroup($groupDn){
+        $users = "No Users Found";
+        $cn = ldap_explode_dn($groupDn,0)[0];
+        $results = ldap_search($this->conn,"OU=Users,OU=People,".$this->basedn, "(memberof:1.2.840.113556.1.4.1941:=$groupDn)",array("cn"));
+        $count = ldap_count_entries($this->conn,$results);
+        if ($count > 0){
+            $users = "<ul>";
+            $entry = ldap_first_entry($this->conn,$results);
+            do {
+                $users .= "<li><span class='text-center label radius' > ".$this->getFirstValue($entry,"cn")."</span></li>";
```
Comment (matthew-lucidchart) [inline core/lucid_ldap.php]: weird to return html from a library

File: core/user.php [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+    public function changePwd($newPwd){
+        $ldapObj = new Lucid_LDAP($this->configFile);
+        $ldapObj -> bind($this->username, $this->password);
+        list ($entry, $dn) = $ldapObj -> searchUser($this -> username,array("sAMAccountName"));
+        $ldapObj -> destroy();
+        $this -> loggerObj -> log( "Changing password from $this->password to $newPwd");
+        $oldPwdEnc = base64_encode(adifyPw($this->password));
+        $newPwdEnc = base64_encode(adifyPw($newPwd));
+        $cmd = "ldapmodify -H $ldapObj->url -D '$dn' -x -w $this->password";
```
Comment (matthew-lucidchart) [inline core/user.php]: escape command line too

File: core/util.php [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+//Returns Client Certificate, PublicKey and PrivateKey for a user of $commonName, $email and length $keyLength
+function generateSslKeypair( $commonName, $mail, $keyLength){
+  $key = openssl_pkey_new(array("private_key_bits" =>$keyLength));
+  $dn = array( 
+     "countryName" => "US",
+     "stateOrProvinceName" => "UT",
+     "localityName" => "South Jordan",
+     "organizationName"      => "Lucidchart",
```
Comment (matthew-lucidchart) [inline core/util.php]: config.

File: forms/admin_chpass.php [hunk 17]
```diff
@@ -0,0 +1,43 @@
+  <form id="adPassForm" action="admin/resetpwd.php" method="POST" data-abide>
```
Comment (matthew-lucidchart) [inline forms/admin_chpass.php]: safer to use absolute urls

File: core/admin.php [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+            "mail" =>  $result["mail"][0],
+            "dn" => $result["distinguishedName"][0]
+        );
+    }
+
+    //Returns the status of deleting attribute $attrib for user given by $dn
+    public function revoke($dn, $attrib){
+        $ldapObj = new Lucid_LDAP($this->configFile);
+        $ldapObj -> bind($this->username, $this->password);
+        $status  = $ldapObj -> delAttribute($dn, $attrib);
```
Comment (matthew-lucidchart) [inline core/admin.php]: verify self != $dn?

Comment (sriraamas) [inline core/admin.php]: Made the above check in admin/revoke.php

**Comment by sriraamas** [comment]
Made changes according to all the comments including XSRF Cookie

File: admin-ss.php [hunk 19]
```diff
@@ -0,0 +1,109 @@
+<!DOCTYPE html>
+<?php
```
Comment (matthew-lucidchart) [inline admin-ss.php]: this should be before any output

File: admin-ss.php [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+setcookie("xsrftoken", "22a7cfaa8a6ba642b3269f32aad8a601", time() + (86400 * 30), "/"); // 86400 = 1 day
```
Comment (matthew-lucidchart) [inline admin-ss.php]: token should be dynamically generated for each user

File: admin/check_username.php [hunk 21]
```diff
@@ -0,0 +1,19 @@
+<?php
+if(!isset($_POST['aname']) || !isset($_POST['apwd']) || !isset($_POST['uname']) || !isset($_COOKIE["xsrftoken"])) {
```
Comment (matthew-lucidchart) [inline admin/check_username.php]: compare cookie xsrf to post xsrf

**Pull request merged**